File: lab5/Dictionary.c
// CMPS 12m Lab5
// Jonathan Montgomery
// jopmontg@ucsc.edu
// DictionaryADT


#include <assert.h>
#include <stdio.h>
#include "Dictionary.h"
#include <stdlib.h>
#include <string.h>

//Struct datatypes based off of IntegerStack.c ADT provided on website

typedef struct NodeField{
  char* key;
  char* value;
  struct NodeField* next;
}NodeField;

typedef NodeField* Node;

typedef struct DictionaryObj{ 
    Node head;
    Node previous;
    int number;
}DictionaryObj;


Node newNode(char* key, char* value){
  Node node = malloc(sizeof(NodeField));
  assert(node!=NULL);
  node->key = key;
  node->value = value;
  node->next = NULL;
  return(node);
}

Dictionary newDictionary(void){
  Dictionary D  = malloc(sizeof(DictionaryObj));
  assert(D!=NULL);
  D->head = NULL;
  D->previous = NULL;
  D->number = 0;
  return D;

}

int isEmpty(Dictionary D){
  if(D==NULL){
    exit(EXIT_FAILURE);
  }
  if(D->number == 0){
    return 1;
  } else {
    return 0;
  }

}

// wrote findkey function but never used
Node findKey(char* k){
  Node index;
  while(index!=NULL){
    if(strcmp(index->key, k)==1){
      return index;
    }else{
      index = index->next;
    }
  }
  return 0;
}

// based on IntegerStack.c function
void freeNode(Node* pN){
  if( pN!=NULL && *pN!=NULL){
    free(*pN);
    *pN = NULL;
  }
}

void freeDictionary(Dictionary* pD){
  free(*pD);
  *pD = NULL;
}

int size(Dictionary D){
  return D->number;
}

// Based on pa3 lookup function
char* lookup(Dictionary D, char* k){
  if(D==NULL){
    exit(EXIT_FAILURE);
  }
  Node temp = D->head;
  while(temp != NULL){ // find key that matches node key and return that value
    if(strcmp(temp->key,k)== 0){
      return temp->value;
    }
    temp = temp->next;
  }
  return 0;
}
void insert(Dictionary D, char* k, char* v){
 if(D==NULL){
   exit(EXIT_FAILURE);
  }
 if(D->head == NULL) { // if no objects, insert will intialize first node
   D->head = D->previous = newNode(k, v);
 } else if(D->number==0){
   D->head = D->previous = newNode(k, v);
 } else { // insert new node, rearrange list pointers
       Node node = newNode(k, v);
       D->previous->next = node;
       D->previous = node;
   }
    D->number++;
}

void delete(Dictionary D, char* k){
  Node temp = D->head;
  if(strcmp(temp->key,k)==0){
    Node temp2 = temp->next;
    D->head = temp2;
    freeNode(&temp);
  }else{
    while(strcmp(temp->next->key, k)!=0){ // check to find location of key and corresponding value
      temp = temp->next;
    }
    temp->next = temp->next->next; // delete by rearranging linked pointers
    freeNode(&temp); // free the node from the list
  }
  D->number--;
}
void makeEmpty(Dictionary D){
  freeNode(&D->head);
  D->number = 0;
}

void printDictionary(FILE* out, Dictionary D){
  for(Node N = D->head; N!=NULL; N = N->next){
    fprintf(out, " %s %s \n ", N->key, N->value); 
  }
}

File: lab7/Dictionary.java
// Cmps12M - Lab 7 
// Implementing BST Dictionary in Java based on C version on website
// Jonathan Montgomery - jopmontg@ucsc.edu


public class Dictionary implements DictionaryInterface{

    private class Node{
	Obj item;
	Node left;
	Node right;

	Node(Obj y){
	  item = y;
	  left = null;
	  right = null;

	}
    }

    private class Obj{
	String key;
	String value;

	Obj(String x, String y){
	    key = x;
	    value = y;
	}
    }

    private Node Root;
    private int Items;

    public Dictionary(){
	Root = null;
	Items = 0;
    }

    private Node findKey(Node Root, String k){
	if(Root == null || Root.item.key.equals(k)){
	    return Root;
	}
	if(Root.item.key.compareTo(k)>0){
	    return findKey(Root.left, k);
	} else {
	    return findKey(Root.right, k);

	}

    }

    private Node findParent(Node New, Node Root){
	Node P = null;
	if(New!=Root){
	    P = Root;
	    while(P.left!=New && P.right!=New){
		if(New.item.key.compareTo(P.item.key)<0){
		    P = P.left;
		}
	     else {
		P = P.right;
	     }
	    }
	}
	return P;
    }

    private Node findLeftmost(Node R){
	Node L = R;
	if(L!=null){
	    for(; L.left!=null; L=L.left){
		return L;
	    }
	}
	return L;
    }

    public boolean isEmpty(){
	return(Items == 0);
    }

    public int size(){
	return Items;
    }

    public String lookup(String key){
	Node N = findKey(Root, key);
	return(N==null ? null : N.item.value);
    }

    public void insert(String key, String value) throws DuplicateKeyException{
	Node A = new Node(new Obj(key, value));
	Node B = Root;
	Node C = null;
	while(B!=null){
	    C = B;
	    if(B.item.key.compareTo(key)>0){
		B = B.left;
	    } else {
		B = B.right;
	    }
	}
    if(C==null){
	Root = A;	
    } else{
	if(C.item.key.compareTo(key)>0){
	    C.left = A;
	}
	else {
	    C.right = A;
	}
      }
       Items++;
    }

    public void delete(String key) throws KeyNotFoundException{
	Node N = findKey(Root, key);
	Node B = findParent(N, Root);
	if(N.left == null && N.right == null){
	    if(N == Root){
		Root = null;
	    } else {
	    Node A = findParent(N, Root);
	    if(A.right == N){
		A.right = null;
	    }
	    else {
		A.left = null;
	    }
	    }
	}
	else if(N.right==null){
		if(N == Root){
		    Root = N.left;
		}
		else{
		    if(B.right == N){
			B.right = N.left;
		    }
		    else {
			B.left = N.left;
		       }
		}
	}
       else if(N.left==null){
	   if(N == Root){
	       Root = N.right;
	   }else{
	       if(B.right == N){
		   B.right = N.right;
	       }else{
		   B.left = N.right;
	       }
	   }
       } else {
	    Node C = findLeftmost(N.right);
	    N.item.key = C.item.key;
	    N.item.value = C.item.value;
	    Node D = findParent(C, N);
	    if(D.right == C){
		D.right = C.right;
	    }
	    else {
		D.left = C.right;
	    }
       }
	  Items--;
    }

    public void makeEmpty(){
	Root = null;
	Items = 0;
    }

    void printInOrder(Node N){
	if(N!=null){
	    printInOrder(N.left);
	    System.out.println(N.item.key + " " + N.item.value);
	    printInOrder(N.right);
	}
    }

    public String toString(){
	printInOrder(Root);
	return " ";
    }
}
